system/lib/libc/musl/src/signal/setitimer.c

```diff
@@ -4,8 +4,78 @@
 
 #define IS32BIT(x) !((x)+0x80000000ULL>>32)
 
+#ifdef __EMSCRIPTEN__
+#include <signal.h>
+#include <stdio.h>
+#include <emscripten/emscripten.h>
+
+// Timeouts can either fire directly from the JS event loop (which calls
+// `_emscripten_timeout`), or from `_emscripten_check_timers` (which is called
+// from `_emscripten_yeild`).  In order to be able to check the timers here we
+// cache the current timeout and intervale for each the 3 types of timer
+// (ITIMER_PROF/ITIMER_VIRTUAL/ITIMER_REAL).
```

Comment:
Does timeout/interval here mean the same as in JS, that is, a timeout is one-shot, and intervals are repeating?

Reply:
Yes, exactly.

Reply:
I'm still quite confused here, then.

First, `setitimer` only sets a one-shot timer, IIUC the C docs. Is that right? But those same docs also mention "interval". So they must mean something else than a repeating timer, but I'm not sure what.

Second, how is this encoded? If say the first of the 3 items here has a timeout, what is the value of the intervals for it? That is, if each item is either a timer or an interval, what is the value of the other set to?

I feel like I'm missing some basic conception of the design here - that's why I'm asking for more docs I guess.

Reply:
Each of the 3 timers is made up of a timeout and an optional interval.   To set a one shot timer you just pass a timeout.  To set a repeating timer you need to set both.  The struct of the timer is C looks like this:

```
           struct itimerval {
               struct timeval it_interval; /* Interval for periodic timer */
               struct timeval it_value;    /* Time until next expiration */
           };
```

When I store them here I convert the timeval struct to a single double for each comparison with emscripten native concept of time as a double.

More from the man page:

```

       These system calls provide access to interval timers, that is, timers that initially
       expire at some point in the future, and  (optionally)  at  regular  intervals  after
       that.
```

Reply:
Thanks, I missed that part about them being repeatable at an interval...